Intro/README.md

```diff
@@ -0,0 +1,455 @@
+we'll ignore algorithm agility.
+
+Other parts of this [tutorial](README.md) may cover specific concepts in
+much more detail.
+
+# Core Concepts
+
+Some core concepts in the world of TPMs
+
+ - cryptography:
+    - digest functions
+    - hash extension
+    - HMAC
+    - symmetric encryption
+    - asymmetric encryption and/or signature algorithms
+ - platform configuration registers (PCRs)
+ - cryptographic object naming
+ - immutability of object public areas
+ - key hierarchies
+ - key wrapping
+ - limited resources
+    - tickets
+    - resource management
+ - sessions
+ - authorization
+    - restricted cryptographic keys
+    - policies
+ - other object types
+    - non-volatile (NV) indexes
+ - attestation
+
+We'll assume reader familiarity with cryptography so we need not explain
+it.
+
+Authorization is the most important aspect of a TPM, since that's
+ultimately what it exists for: to authorize a system or application to
+perform certain duties when all the desired conditions allow for it.
+
+TPMs have a very rich set of options for authorization.  It's not just
+[policies](#Policies), but also cryptographic object names used with
+restricted keys to allow access only to applications that also have
+other access.
+
+Where to start?  Let's start with hash extension, which may be the only
+trivial concept in the world of TPMs!
+
+## Important Notes
+
+ - Chained implications.
+
+   It is important to note the TPM ecosystem's critical dependence on
+   TPMs performing authorization as specified.  A lot can be
+   accomplished by chaining cryptographic implications.  Sending a
+   secret to a TPM-using application, with the secret encrypted in that
+   TPM's primary endorsement public key is safe when one knows that
+   public key truly is a legitimate hardware TPM's public key: the TPM
+   will not release the secret to the application unless the recipient
+   has some access specified by the sender.  Complex chains of events
+   can be secure provided one knows that it involves legitimate hardware
+   TPMs that otherwise could not be secured.  Therefore the legitimacy
+   of a TPM that one communicates with is essential.
+
+ - Limited resources.
+
+   Hardware TPMs are very resource-starved.  The set of commands defined
+   by the TCG TPM specifications is specifically constructed to a) be
+   small and simple (simple-enough anyways), yet b) enable rich
+   semantics in spite of this.
+
+   These limited resources mean, among other things, that TPMs don't do
+   complex things like PKIX certificate validation path construction, or
+   even PKIX certificate path validation, nor do they implement complex
+   authentication protocols using PKIX, nor Kerberos, nor even simple
+   protocols like JWT.  Hardware TPMs can be used to implement portions
+   of such complex protocols (e.g., certificate issuance), but they
+   cannot implement the entirety of such protocols.
+
+ - Hardware TPMs are _slow_.
+
```

Comment:
They don't need to be, but the spec says so for security reasons.

Reply:
Software vTPMs aren't slow.  The issue is lack of resources -- HW TPMs have to be _cheap_.

Reply:
If the spec says they have to be slow, I've missed it.

Reply:
I have also missed mentioning of that in the spec. In any case, the TPM is designed for security, therefore a trade-off with performance is made. In contrast, modern x86 CPUs have speculative branching and execution to trade performance for security (the meltdown saga) ;)

Reply:
" Hardware TPMs are _slow_."  Another opinion.  Slow compared to what.  It would be more useful to just give performance numbers and let the reader decide.

Reply:
Compared to the CPU.

Reply:
I didn't think there would be any controversy about "HW TPMs are slow"...

Reply:
I've removed that section altogether, @kgold2, but https://github.com/wolfSSL/wolfTPM does have benchmarks, and... HW TPMs do look pretty slow to me by comparison to CPUs and, therefore presumably TPM simulators running on CPUs.